controllers/external/tracker.go

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package external
+
+import (
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// ObjectTracker is a helper struct to deal when watching external unstructured objects.
+type ObjectTracker struct {
+	m sync.Map
+
+	Controller controller.Controller
+}
+
+// Watch uses the controller to issue a Watch only if the object hasn't been seen before.
+func (o *ObjectTracker) Watch(log logr.Logger, obj runtime.Object, handler handler.EventHandler) error {
+	gk := obj.GetObjectKind().GroupVersionKind().GroupKind()
+
+	_, loaded := o.m.LoadOrStore(gk.String(), struct{}{})
+	if loaded || o.Controller == nil {
```

Comment:
I find `o.Controller == nil` test a little bit confusing here.
What about moving the test at the beginning of the func and returning error vs return nil

Reply:
Fixed, thanks!